packages/live-status-gateway/src/wsHandler.ts

```diff
@@ -71,108 +99,226 @@ export interface WebSocketTopic {
+		this._observers.set(callback, { keysToPick, lastData: this.shallowClone(this._collectionData) })
 	}
 
-	async unsubscribe(observer: CollectionObserver<T>): Promise<void> {
-		this._logger.info(`${observer.observerName}' removed observer for '${this._name}'`)
-		this._observers.delete(observer)
+	/**
+	 * Called after a batch of updates to documents in the collection
+	 */
+	protected changed(): void {
+		// override me
 	}
 
-	async notify(data: T | undefined): Promise<void> {
-		for (const observer of this._observers) {
-			await observer.update(this._name, data)
+	notify(data: T | undefined): void {
+		for (const [observer, o] of this._observers) {
+			if (
+				!o.lastData ||
+				!o.keysToPick ||
+				!data ||
+				!arePropertiesShallowEqual(o.lastData, data, undefined, o.keysToPick)
+			) {
+				observer(data)
+				o.lastData = this.shallowClone(data)
+			}
 		}
 	}
 
+	protected shallowClone(data: T | undefined): T | undefined {
+		if (data === undefined) return undefined
+		if (Array.isArray(data)) return [...data] as T
+		if (typeof data === 'object') return { ...data }
+		return data
+	}
+
 	protected logDocumentChange(documentId: string | ProtectedString<any>, changeType: string): void {
 		this._logger.silly(`${this._name} ${changeType} ${documentId}`)
 	}
 
 	protected logUpdateReceived(collectionName: string, updateCount: number | undefined): void
-	protected logUpdateReceived(collectionName: string, source: string, extraInfo?: string): void
+	protected logUpdateReceived(collectionName: string, extraInfo?: string): void
 	protected logUpdateReceived(
 		collectionName: string,
-		sourceOrUpdateCount: string | number | undefined,
-		extraInfo?: string
+		extraInfoOrUpdateCount: string | number | undefined | null = null
 	): void {
-		if (typeof sourceOrUpdateCount === 'string') {
-			let message = `${this._name} received ${collectionName} update from ${sourceOrUpdateCount}`
-			if (extraInfo) {
-				message += `, ${extraInfo}`
-			}
-			this._logger.debug(message)
-		} else {
-			this._logger.debug(`'${this._name}' handler received ${sourceOrUpdateCount} ${collectionName}`)
+		let message = `${this._name} received ${collectionName} update`
+		if (typeof extraInfoOrUpdateCount === 'string') {
+			message += `, ${extraInfoOrUpdateCount}`
+		} else if (extraInfoOrUpdateCount !== null) {
+			message += `(${extraInfoOrUpdateCount})`
 		}
+		this._logger.debug(message)
 	}
 
 	protected logNotifyingUpdate(updateCount: number | undefined): void {
 		this._logger.debug(`${this._name} notifying update with ${updateCount} ${this._collectionName}`)
 	}
+
+	protected getCollectionOrFail(): CoreCollection<CollectionDocCheck<CorelibPubSubCollections[TCollection]>> {
+		const collection = this._core.getCollection<TCollection>(this._collectionName)
+		if (!collection) throw new Error(`collection '${this._collectionName}' not found!`)
+		return collection
+	}
+}
+
+export abstract class PublicationCollection<
```

Comment:
Do yo think it would make sense to put thse abstract classes to files of their own? I suppose they aren't really related to the `WebSocketTopicBase` as such?

Reply:
I extracted those classes as advised